src/gromacs/nblib/topology.cpp

```diff
@@ -188,13 +189,22 @@ std::vector<real> TopologyBuilder::extractAtomTypeQuantity(Extractor extractor)
 Topology TopologyBuilder::buildTopology()
 {
     topology_.excls_  = createExclusionsList();
-    topology_.masses_ = extractAtomTypeQuantity(
+    topology_.masses_ = extractAtomTypeQuantity<real>(
             [](const auto& data, auto& map) { return map[data.atomTypeName_].mass(); });
-    topology_.charges_ = extractAtomTypeQuantity([](const auto& data, auto& map) {
+    topology_.charges_ = extractAtomTypeQuantity<real>([](const auto& data, auto& map) {
         ignore_unused(map);
         return data.charge_;
     });
 
+    topology_.atomTypes_ = extractAtomTypeQuantity<std::string>(
+            [](const auto& data, auto& map) { return map[data.atomTypeName_].name(); });
+    //! Create a temporary list to sort the AtomType names and remove duplicates
+    std::list<std::string> atomTypesList(std::begin(topology_.atomTypes_), std::end(topology_.atomTypes_));
```

Comment:
We will also need to do this for the C6 and C12 parameters so it it probably a good idea to extract this to a function.

Reply:
Would you like to do this within the current PR?

Reply:
Would be nice to have reusable functions as opposed to stray lambdas essentially repeating the same logic in the code.

Reply:
I think that this change is large enough that extracting this should be done in a follow-up patch, unless there is a concern that this impairs readability or correctness.

Reply:
Sounds good!

Reply:
Sure! This can be pushed to a follow-up patch.

Reply:
What do you want to extract into a function, i.e. what is this? Just the filling of the vectors for c6 and c12?
or do you also need both of them sorted and unique?

Reply:
Also, in terms of API design, this is an issue. API's are supposed to be as succinct as possible and this
here adds another function to it that essentially just does sorting and remove duplicates. Sorting
and uniqueness are completely unrelated to this class and we can provide generation of a sorted unique sequence as a free function. Or - if we would have to call this multiple times it could be part of the private interface.

Reply:
I made another PR that shows what (I think) we need for setting up the nonbonded parameters. I agree that this design is not such a good one. I just wanted to have something to motivate what is needed.

Reply:
There needs to be a vector<int> of atom types that is used with the nonbonded params to actually create the gromacs structures that are used for nonbonded force calculation.